r/R/convert-array-stream.R

```diff
@@ -62,26 +68,52 @@ convert_array_stream <- function(array_stream, to = NULL, size = NULL, n = Inf)
     )
   }
 
+  # Collect all batches into a list()
+  batches <- collect_array_stream(
+    array_stream,
+    n,
+    schema = schema,
+    validate = FALSE
+  )
+
+  # If there is exactly one batch, use convert_array()
+  if (length(batches) == 1L) {
+    return(.Call(nanoarrow_c_convert_array, batches[[1]], to))
+  }
+
+  # Otherwise, compute the final size, create another array stream,
+  # and call convert_array_stream() with a known size.
+  lengths <- vapply(
+    batches,
+    # Use custom accessor because array$length in a loop is slow
+    function(array) {
+      .Call(nanoarrow_c_array_proxy, array, NULL, FALSE)$length
+    },
+    double(1)
+  )
+  basic_stream <- .Call(nanoarrow_c_basic_array_stream, batches, schema, FALSE)
+  convert_array_stream(basic_stream, to = to, size = sum(lengths))
```

Comment:
Maybe something like this, to avoid the tail recursion and extra work?

```suggestion
  batches <- collect_array_stream(
    array_stream,
    n,
    schema = schema,
    validate = FALSE
  )

  # We are guaranteed to have only one batch here:
  stopifnot(length(batches) == 1)
  .Call(nanoarrow_c_convert_array, batches[[1]], to)
```